src/modules/types/ContestTypeHelper.ts

```diff
@@ -0,0 +1,71 @@
+
+export default class ContestTypeHelper {
+    public static contestTypeMatrix: Array<Array<number>> = (() => {
+        const imm = TypeEffectivenessValue.Immune;
+        const not = TypeEffectivenessValue.NotVery;
+        const neu = TypeEffectivenessValue.Neutral;
+        return [
+            //       B                   B
+            //       E                   A
+            //       A                   L
+            //       U                   A
+            //       T         S    T    C      <- Defending type
+            //  C    I    C    M    O    N    N
+            //  O    F    U    A    U    C    O   Attack type
+            //  O    U    T    R    G    E    N        |
+            //  L    L    E    T    H    D    E        v
+            [neu, not, imm, imm, not, not, imm], // COOL
+            [not, neu, not, imm, imm, not, imm], // BEAUTIFUL
+            [imm, not, neu, not, imm, not, imm], // CUTE
+            [imm, imm, not, neu, not, not, imm], // SMART
+            [not, imm, imm, not, neu, not, imm], // TOUGH
+            [not, not, not, not, not, not, imm], // BALANCED
+            [imm, imm, imm, imm, imm, imm, imm], // NONE
+        ];
+    })();
+
+    public static getAppealModifier(
+        a1: ContestType = ContestType.None,
+        a2: ContestType = ContestType.None,
+        a3: ContestType = ContestType.None,
+        d1: ContestType = ContestType.None,
+        d2: ContestType = ContestType.None,
+        d3: ContestType = ContestType.None,
+    ): number {
+        if (a1 === ContestType.None || d1 === ContestType.None) {
+            return 1;
+        }
+
+        // Apply None type where there isn't an attacking type
+        // eslint-disable-next-line no-param-reassign
+        a2 = a2 !== ContestType.None ? a2 : 6;
+        // eslint-disable-next-line no-param-reassign
+        a3 = a3 !== ContestType.None ? a3 : 6;
+        // eslint-disable-next-line no-param-reassign
+        d2 = d2 !== ContestType.None ? d2 : 6;
+        // eslint-disable-next-line no-param-reassign
+        d3 = d3 !== ContestType.None ? d3 : 6;
+
+        let m1 = ContestTypeHelper.contestTypeMatrix[a1][d1];
+        let m2 = ContestTypeHelper.contestTypeMatrix[a2][d1];
+        let m3 = ContestTypeHelper.contestTypeMatrix[a3][d1];
+        
+        let m4 = ContestTypeHelper.contestTypeMatrix[a1][d2];
+        let m5 = ContestTypeHelper.contestTypeMatrix[a2][d2];
+        let m6 = ContestTypeHelper.contestTypeMatrix[a3][d2];
+
+        let m7 = ContestTypeHelper.contestTypeMatrix[a1][d3];
+        let m8 = ContestTypeHelper.contestTypeMatrix[a2][d3];
+        let m9 = ContestTypeHelper.contestTypeMatrix[a3][d3];
+
+        // Find the numbers for each type, cap at 1
+        let type1 = Math.min(m1 + m2 + m3, 1);
+        let type2 = Math.min(m4 + m5 + m6, 1);
+        let type3 = Math.min(m7 + m8 + m9, 1);
+
+        // Find the most effective type and use it
+        return Math.max(type1, type2, type3);
```

Comment:
Type balancing was a bit troublesome. The intention is to be different from TypeHelper multipliers.

I approached this with the idea of audience appeal in mind, so if a pokemon is able to "max" appeal, I want that to be registered over an immunity. Ex. If a pokemon in a Cool contest is Cool and Cute, the audience won't dislike it for being Cute, because they only care if it's being Cool, therefore the closer it is to Cool (which is where adjacent types like Tough and Beautiful in this case come in), the more "points" it gets. I dabbled with the idea of different calculations for audience and opponent pokemon, but that just made everything too messy, so I prioritized the audience part.

The three type sweet spot is also intentional, although I'm a bit less certain on whether it's the right move. The idea here is to give incentive to feed type-adding pokeblocks to pokemon. Only a handful of pokemon have naturally-occurring triple types... but there's nothing stopping you from putting all your regular blocks in just them. I have always envisioned ribbons to help encourage feeding multiple pokemon, but it would be preferable to have a gameplay element do that too somehow. Of course, it would also be possible to redo the types completely, and even remove `Balanced`, if you're up to some extreme modifications. I kind of felt my way through with the current type chart iteration, so I wouldn't be surprised if there's something better out there.

Reply:
This is all great feedback and I'm really grateful you're getting into the nitty gritty for me. Game mechanics are always a tough one. Let me lay down my intentions with the ranks and bars first, both to get feedback from you, and because I think better by writing out my thoughts.
Contests are divided into four sets, each with their own gimmick:
- Hoenn's `Normal` - `Master`: no gimmick, introductory
- Sinnoh's `Super Normal` - `Super Master`: some gimmick
- Hoenn's `Spectacular`: different gimmick from regular Sinnoh set
- Sinnoh's `Brilliant Shining`: have regular Sinnoh and `Spectacular` gimmicks come together

Within each contest itself, we have two things going on:
- Audience: this determines if you've won the contest or not
  - These take on the contest type of the contest that is selected, so the points dealt to the bar in each contest is different
- Trainers: this is meant to be the main way of gaining Contest Tokens, hence the Trainer Streak

So we have, in theory, a way to influence damage by rank and contest type, and we are left with audience and trainer bars that are affected by a similar formula. You mentioned some good ideas for what I'm assuming is the trainer bar? Here:
> Maybe if the opponent's pokemon matches the contest type, it always defends with that type and gets a bonus, but if not your pokemon get to attack its weakest type.

You'll have to elaborate on what you mean by "always defends with that type", but I like the idea of a bonus and further connecting the opponent with the contest's `ContestType`. If it's done well enough I can even scrap the bonus upon defeating it.

You also mentioned a battlefield effect. `Spectacular` rank's type-restricting gimmick is the closest to that, but is too extreme to apply across all ranks, and it still has the problem of the audience bar increasing with the same formula as the trainer bar. This is where the idea of Sheen comes into play. Sheen could influence how much attack is being dealt to the opponent. There could be a debuff affecting each rank's opponents, and while a pokemon who's type matches the contest's wouldn't need sheen to attack the opponent with 100% of its appeal, other pokemon would need to offset the debuff with a high enough sheen. I'm unsure if this overcomplicates things, and also if it should exist for the introductory Hoenn contests or not, but it's something I just came up with and curious to see if it would lead to somewhere.